util/retrieve-cluster-logs.py

```diff
@@ -0,0 +1,213 @@
+        "--bucket-prefix", help="Keypath under which exported CloudWatch logs data will be stored in s3 bucket."
+    )
+    parser.add_argument("--region", required=True, help="Region in which the CloudWatch log group exists.")
+    args = parser.parse_args()
+    if not args.bucket_prefix:
+        args.bucket_prefix = get_default_bucket_prefix(args.cluster)
+    args.log_group = get_log_group(args)
+    verify_bucket_exists_in_region(args)
+    return args
+
+
+def start_export_task(logs_client, args):
+    """Start the task that will export the cluster's logs to an s3 bucket, and return the task ID."""
+    LOGGER.info(
+        "Starting export of logs from log group {log_group} to s3 bucket {bucket}".format(
+            log_group=args.log_group.get("logGroupName"), bucket=args.bucket
+        )
+    )
+    start_time = datetime.fromtimestamp(args.log_group.get("creationTime") / 1000)
+    end_time = datetime.now()
+    try:
+        response = logs_client.create_export_task(
+            logGroupName=args.log_group.get("logGroupName"),
+            fromTime=int(1000 * start_time.timestamp()),
+            to=int(1000 * end_time.timestamp()),
+            destination=args.bucket,
+            destinationPrefix=args.bucket_prefix,
+        )
+    except Exception as err:
+        if "Please check if CloudWatch Logs has been granted permission to perform this operation." in str(err):
+            err_and_exit(
+                "CloudWatch Logs needs GetBucketAcl and PutObject permisson for the s3 bucket {bucket}. See {url} for "
+                "more details.".format(
+                    bucket=args.bucket,
+                    url="https://docs.aws.amazon.com/AmazonCloudWatch/latest/logs/S3ExportTasks.html#S3Permissions",
+                )
+            )
+        else:
+            err_and_exit("Unexpected error when starting export task: {}".format(err))
+    return response.get("taskId")
+
+
+def get_status_for_export_task(logs_client, task_id):
+    """Get the status for the CloudWatch export task with the given task_id."""
+    tasks = logs_client.describe_export_tasks(taskId=task_id).get("exportTasks")
+    if not tasks:
+        err_and_exit("Unable to get status for CloudWatch logs export task with ID={}".format(task_id))
+    elif len(tasks) > 2:
+        err_and_exit(
+            "More than one CloudWatch logs export task with ID={task_id}:\n{tasks}".format(
+                task_id=task_id, tasks=json.dumps(tasks, indent=2)
+            )
+        )
+    return tasks[0].get("status").get("code")
+
+
+def wait_for_task_completion(logs_client, task_id):
+    """Wait for the CloudWatch logs export task given by task_id to finish."""
+    LOGGER.info("Waiting for export task with task ID {} to finish.".format(task_id))
+    status = "PENDING"
+    still_running_statuses = ("PENDING", "PENDING_CANCEL", "RUNNING")
+    while status in still_running_statuses:
+        status = get_status_for_export_task(logs_client, task_id)
```

Comment:
we probably want to set a sleep between consecutive calls.

Reply:
Makes sense. I'll set it to sleep for one second in between calls.